### PR TITLE
[Block Editor]: Fix content loss from `replaceInnerBlocks` with controlled blocks

### LINKED ISSUE #41887: Post content deleted when adding column to template
### Description

When editing a post template in the single template editor by wrapping the Post Content block inside a Columns block, the content of Post Content gets deleted (just the content, not the block itself).

### Step-by-step reproduction instructions

1. Add content to a post, save and publish.
2. In the post sidebar, go to Template section and create a new template
3. Locate the Group block that contains Post Content, and wrap it in a Columns block.
4. Add a second column to the Columns block.
5. Save and view post; verify the content is not longer there

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Fixes: https://github.com/WordPress/gutenberg/issues/41887

When we dispatch `REPLACE_INNER_BLOCKS` action we have a special handling which consists of two steps:
1. Remove blocks
2. Insert blocks

In this flow(`withReplaceInnerBlocks`) we have some special handling for controlled blocks(reusable, postContent, etc..) where we keep them in state during the removal(`keepControlledInnerBlocks`) and we are updating the `order` again after insertion.

What we are missing is the update of `block tree`. I think the only reason we missed that until now is because in our code base we have just a few calls to `replaceInnerBlocks`, where most of them(if not all) aren't affected by this bug - example use cases would be replacing blocks with entirely new blocks or replacing blocks that only specific blocks are allowed(non controlled blocks).

## Testing Instructions(also can be seen in the video)
1. In a page add some content
2. Go to `edit` the template
3. Transform the `PostContent` to `Columns` and add a new columns(this is where the call to `replaceInnerBlocks` happens).
4. If you save, you'll observe that now you've updated only the template and the content is preserved.

#### Before

https://user-images.githubusercontent.com/16275880/175545154-67da1d8c-bb9f-4dc8-9721-97e0d17a7104.mov

#### After

https://user-images.githubusercontent.com/16275880/175545168-4e1517de-df3e-4b01-b949-5ea4832f2dd2.mov

## Notes
1. **Good testing is needed here to avoid any regression**, as content loss is really really bad 😄 . Please test extensively.
2. I think this is the best place to put this code, but the whole logic is very complex, so I'd need more eyes on this one!
3. For `remove blocks and replace blocks` with have similar handling in a [different flow](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/store/reducer.js#L322)(`REPLACE_BLOCKS_AUGMENTED_WITH_CHILDREN`, `REMOVE_BLOCKS_AUGMENTED_WITH_CHILDREN `)
4. `controlled` blocks are the blocks that are a separate entity like `reusable blocks, template parts, etc..`.

### CODE REVIEW

File: packages/block-editor/src/store/reducer.js [hunk 1]
```diff
@@ -689,7 +689,8 @@ const withReplaceInnerBlocks = ( reducer ) => ( state, action ) => {
 			index: 0,
 		} );
 
-		// We need to re-attach the block order of the controlled inner blocks.
+		// We need to re-attach the controlled inner blocks to the blocks tree and
+		// preserve their block order.
```
Comment (mtias) [inline packages/block-editor/src/store/reducer.js]: Can we remove the line break between the sentences? Reads a bit awkward. Also there are two periods after entity to fix :)

**Review by tellthemachines — APPROVED** [review]
Thanks for the prompt fix! Testing with the original reproduction steps everything now works as expected. I also tested adding/removing columns from a Columns block in a regular post, with both a Navigation block and a reusable block inside, and no content is lost. (On trunk, the same flow results in the content of both the Navigation and reusable block disappearing.)

Offhand I can't think how updating the state tree here might cause any harm, but I don't know this part of the code in-depth, so you might want to double check that 😅

**Review by talldan — COMMENTED** [review]
Great work figuring this out. In `trunk` the tree is definitely computed incorrectly, and in this branch I couldn't find a situation where it's incorrect.

I had quite a deep dive into the code to try to understand it. In terms of the solution, I find it difficult to say whether the controlled block should be reattached to the tree as it is in this PR, or whether it shouldn't be removed in the first place. To explain what I mean, as part of replacing the inner blocks, the blocks are first removed and then reinserted. There's a `keepControlledInnerBlocks` property of the `REMOVE_BLOCKS` action:
https://github.com/WordPress/gutenberg/blob/f6f0cea531867665852d0bb8c4206c16703c8461/packages/block-editor/src/store/reducer.js#L678-L682

But the `withBlockTree` higher order reducer doesn't seem to respect that prop and removes those clientIds anyway:
https://github.com/WordPress/gutenberg/blob/f6f0cea531867665852d0bb8c4206c16703c8461/packages/block-editor/src/store/reducer.js#L383-L395

An alternative solution might be just to keep them in the root, but I don't know for sure—maybe that would have some incorrect consequences.

Whatever the solution, it'd be good to add some unit tests to this PR if it's not too difficult.

**Comment by ntsekouras** [comment]
Thanks for the reviews!

>I find it difficult to say whether the controlled block should be reattached to the tree as it is in this PR, or whether it shouldn't be removed in the first place.

That is something I wasn't sure as well, but all these flows are interconnected with multiple actions and also events in `useBlockSync` which is quite complex and hard to follow and understand(at least for me). My first instinct was to update the state in the respective reducers, but we have very specific handling for [REPLACE_INNER_BLOCKS](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/store/reducer.js#L651) and might be a strong reason for this, so I kept it in there with the other state update(`order`). I'm not in the place yet to know if we can make this better and how, with knowing all the affected places.

I will see to write some tests..

**Review by mcsf — APPROVED** [review]
This looks good to my untrained eye, based on the knowledge I could brush up around controlled blocks. Thanks for the comprehensive test, too.

I haven't done any comprehensive testing, so more than anything I trust your judgment.
